Code/Engine/GameEngine/StateMachine/Implementation/StateMachine.cpp

```diff
@@ -0,0 +1,462 @@
+
+ezResult ezStateMachineDescription::Serialize(ezStreamWriter& originalStream) const
+{
+  originalStream.WriteVersion(s_StateMachineDescriptionVersion);
+
+  ezStringDeduplicationWriteContext stringDeduplicationWriteContext(originalStream);
+  ezTypeVersionWriteContext typeVersionWriteContext;
+  auto& stream = typeVersionWriteContext.Begin(stringDeduplicationWriteContext.Begin());
+
+  ezUInt32 uiNumTransitions = m_FromAnyTransitions.GetCount();
+
+  // states
+  {
+    const ezUInt32 uiNumStates = m_States.GetCount();
+    stream << uiNumStates;
+
+    for (auto& stateContext : m_States)
+    {
+      auto pStateType = stateContext.m_pState->GetDynamicRTTI();
+      typeVersionWriteContext.AddType(pStateType);
+
+      stream << pStateType->GetTypeName();
+      EZ_SUCCEED_OR_RETURN(stateContext.m_pState->Serialize(stream));
+
+      uiNumTransitions += stateContext.m_Transitions.GetCount();
+    }
+  }
+
+  // transitions
+  {
+    stream << uiNumTransitions;
+
+    auto SerializeTransitions = [&](const TransitionArray& transitions, ezUInt32 uiFromStateIndex) -> ezResult {
+      for (auto& transitionContext : transitions)
+      {
+        const ezUInt32 uiToStateIndex = transitionContext.m_uiToStateIndex;
+
+        stream << uiFromStateIndex;
+        stream << uiToStateIndex;
+
+        auto pTransitionType = transitionContext.m_pTransition->GetDynamicRTTI();
+        typeVersionWriteContext.AddType(pTransitionType);
+
+        stream << pTransitionType->GetTypeName();
+        EZ_SUCCEED_OR_RETURN(transitionContext.m_pTransition->Serialize(stream));
+      }
+
+      return EZ_SUCCESS;
+    };
+
+    EZ_SUCCEED_OR_RETURN(SerializeTransitions(m_FromAnyTransitions, ezInvalidIndex));
+
+    for (ezUInt32 uiFromStateIndex = 0; uiFromStateIndex < m_States.GetCount(); ++uiFromStateIndex)
+    {
+      auto& transitions = m_States[uiFromStateIndex].m_Transitions;
+
+      EZ_SUCCEED_OR_RETURN(SerializeTransitions(transitions, uiFromStateIndex));
+    }
+  }
+
+  EZ_SUCCEED_OR_RETURN(typeVersionWriteContext.End());
+  EZ_SUCCEED_OR_RETURN(stringDeduplicationWriteContext.End());
+
+  return EZ_SUCCESS;
+}
+
+ezResult ezStateMachineDescription::Deserialize(ezStreamReader& stream)
+{
+  const auto uiVersion = stream.ReadVersion(s_StateMachineDescriptionVersion);
+
+  ezStringDeduplicationReadContext stringDeduplicationReadContext(stream);
+  ezTypeVersionReadContext typeVersionReadContext(stream);
+
+  ezStringBuilder sTypeName;
+
+  // states
+  {
+    ezUInt32 uiNumStates = 0;
+    stream >> uiNumStates;
+
+    for (ezUInt32 i = 0; i < uiNumStates; ++i)
+    {
+      stream >> sTypeName;
+      if (const ezRTTI* pType = ezRTTI::FindTypeByName(sTypeName))
+      {
+        ezUniquePtr<ezStateMachineState> pState = pType->GetAllocator()->Allocate<ezStateMachineState>();
+        EZ_SUCCEED_OR_RETURN(pState->Deserialize(stream));
+
+        EZ_VERIFY(AddState(std::move(pState)) == i, "Implementation error");
```

Comment:
Instead of EZ_VERIFY you can just use ".AssertSuccess()" on ezResult.

Reply:
AddState doesn't return an ezResult and I need to compare to i